File: data-code/H3_HCRIS_PPS.R

# Meta --------------------------------------------------------------------

## Notes:  PPS (Prospective Payment System) minimum dataset from NBER
##         These data use the same underlying Form 2552 as HCRIS but are
##         pre-extracted flat files covering 1985-1999.
##         - 1985-1995: limited variable set (no balance sheet, no revenue stmt)
##         - 1996-1999: full variable set including balance sheet and revenue stmt
##         - 1996-1999 files are split into two parts due to Stata variable limits
##         Date formats:
##         - Pre-1996: Julian YYDDD (e.g., "91274" = Oct 1, 1991)
##         - 1996-1999: CCYYMMDD (e.g., "19971001")
##         Output dates formatted as M/D/YYYY to match HCRIS H1/H2 output.


# SSA state code mapping --------------------------------------------------

ssa.state <- read_csv('data-code/pps-xwalk/ssa_state_codes.csv',
                       col_types=cols(ssa_code=col_character(), state=col_character()))


# Helper functions ---------------------------------------------------------

## Derive state from provider number (first 2 digits = SSA state code)
get_state_from_provider <- function(provider_num) {
  ssa_code <- substr(as.character(provider_num), 1, 2)
  state_map <- setNames(ssa.state$state, ssa.state$ssa_code)
  return(unname(state_map[ssa_code]))
}

## Pad provider number to 6 digits (preserve leading zeros)
pad_provider <- function(provider_num) {
  provider_num <- trimws(as.character(provider_num))
  ifelse(is.na(provider_num) | provider_num == "", NA_character_,
         str_pad(provider_num, width = 6, side = "left", pad = "0"))
}

## Parse Julian YYDDD date strings to M/D/YYYY
parse_julian_date <- function(date_str) {
  date_str <- trimws(as.character(date_str))
  result <- rep(NA_character_, length(date_str))
  valid <- !is.na(date_str) & nchar(date_str) >= 5
  if (any(valid)) {
    yy <- as.integer(substr(date_str[valid], 1, 2))
    ddd <- as.integer(substr(date_str[valid], 3, 5))
    year <- ifelse(yy >= 80, 1900 + yy, 2000 + yy)
    dates <- as.Date(paste0(year, "-01-01")) + (ddd - 1)
    result[valid] <- format(dates, "%m/%d/%Y")
  }
  return(result)
}

## Parse CCYYMMDD date strings to M/D/YYYY
parse_ccyymmdd <- function(date_str) {
  date_str <- trimws(as.character(date_str))
  result <- rep(NA_character_, length(date_str))
  valid <- !is.na(date_str) & nchar(date_str) == 8
  if (any(valid)) {
    dates <- ymd(date_str[valid])
    result[valid] <- format(dates, "%m/%d/%Y")
  }
  return(result)
}


# Import PPS 1996-1999 (Form 2552-96) -------------------------------------
## These files have the full variable set including balance sheet

for (i in 1996:1999) {

  ## file naming: pps96, pps97, pps98, pps99
  yy <- substr(as.character(i), 3, 4)

  ## load and merge split files
  pps.part1 <- haven::read_dta(paste0("data/input/hcris_pps_nber/pps", yy, "_f1_f1800.dta"))
  pps.part2 <- haven::read_dta(paste0("data/input/hcris_pps_nber/pps", yy, "_f1801_to_end.dta"))
  pps.part1 <- pps.part1 %>% mutate(f1 = pad_provider(f1))
  pps.part2 <- pps.part2 %>% mutate(f1 = pad_provider(f1))
  pps.data <- left_join(pps.part1, pps.part2, by="f1")

  ## extract and rename variables per crosswalk
  final.reports <- pps.data %>%
    transmute(
      report = NA_character_,
      provider_number = f1,
      npi = NA_character_,
      fy_start = parse_ccyymmdd(f22),
      fy_end = parse_ccyymmdd(f23),
      date_processed = NA_character_,
      date_created = parse_ccyymmdd(f69),
      status = as.character(f74),
      year = i,
      data_source = "pps",

      ## facility characteristics
      beds = as.numeric(f85),

      ## charges and revenue (Worksheet G-3)
      tot_charges = as.numeric(f2135),
      net_pat_rev = as.numeric(f2137),
      tot_discounts = as.numeric(f2136),
      tot_operating_exp = as.numeric(f2138),

      ## inpatient detail
      ip_charges = as.numeric(f2133),
      icu_charges = as.numeric(f2108),
      ancillary_charges = as.numeric(f2115),

      ## discharges
      tot_discharges = as.numeric(f178),
      mcare_discharges = as.numeric(f172),
      mcaid_discharges = as.numeric(f175),

      ## Medicare payments (Worksheet E Part A)
      tot_mcare_payment = as.numeric(f1833),
      secondary_mcare_payment = as.numeric(f1834),

      ## address info
      street = NA_character_,
      city = NA_character_,
      state = get_state_from_provider(f1),
      zip = NA_character_,
      county = NA_character_,
      name = as.character(f18),

      ## uncompensated care and cost-to-charge
      uncomp_care = NA_real_,
      cost_to_charge = NA_real_,

      ## capital and assets
      new_cap_ass = as.numeric(f1473),
      cash = as.numeric(f2038),
      fixed_assets = as.numeric(f2066),

      ## depreciation (accumulated, from balance sheet)
      depr_land = as.numeric(f2050),
      depr_bldg = as.numeric(f2054),
      depr_lease = as.numeric(f2056),
      depr_fixed_equip = as.numeric(f2058),
      depr_auto = as.numeric(f2060),
      depr_major_equip = as.numeric(f2062),
      depr_minor_equip = as.numeric(f2064),

      ## balance sheet
      current_assets = as.numeric(f2048),
      current_liabilities = as.numeric(f2081),

      ## PPS-equivalent variables (only populated in v1996 HCRIS)
      pps_ip_charges = NA_real_,
      pps_op_charges = NA_real_,
      pps_mcare_cost = NA_real_,
      pps_pgm_cost = NA_real_
    )

  assign(paste0("final.reports.", i), final.reports)
  if (i == 1996) {
    final.hcris.pps96 <- final.reports
  } else {
    final.hcris.pps96 <- rbind(final.hcris.pps96, final.reports)
  }
}


# Import PPS 1985-1995 (Forms 2552-85, 2552-89, 2552-92) ------------------
## These files have a more limited variable set
## No balance sheet, no revenue/expense statement, no capital asset detail
## Dates are in Julian YYDDD format

for (i in 1985:1995) {

  yy <- substr(as.character(i), 3, 4)
  pps.data <- haven::read_dta(paste0("data/input/hcris_pps_nber/pps", yy, ".dta")) %>%
    mutate(provno = pad_provider(provno))

  final.reports <- pps.data %>%
    transmute(
      report = NA_character_,
      provider_number = provno,
      npi = NA_character_,
      fy_start = parse_julian_date(begdate),
      fy_end = parse_julian_date(enddate),
      date_processed = NA_character_,
      date_created = NA_character_,
      status = as.character(status),
      year = i,
      data_source = "pps",

      ## facility characteristics
      beds = as.numeric(totbeds),

      ## charges (from cost allocation worksheets, not revenue stmt)
      tot_charges = as.numeric(totalg),
      net_pat_rev = NA_real_,
      tot_discounts = NA_real_,
      tot_operating_exp = as.numeric(opertots),

      ## inpatient detail
      ip_charges = as.numeric(mtotalg),
      icu_charges = NA_real_,
      ancillary_charges = as.numeric(outpatg),

      ## discharges
      tot_discharges = as.numeric(f88),
      mcare_discharges = as.numeric(f82),
      mcaid_discharges = as.numeric(f84),

      ## Medicare payments
      tot_mcare_payment = as.numeric(amtdue),
      secondary_mcare_payment = NA_real_,

      ## address info
      street = NA_character_,
      city = NA_character_,
      state = get_state_from_provider(provno),
      zip = NA_character_,
      county = NA_character_,
      name = as.character(provname),

      ## uncompensated care and cost-to-charge
      uncomp_care = NA_real_,
      cost_to_charge = NA_real_,

      ## capital and assets (not available pre-1996)
      new_cap_ass = NA_real_,
      cash = NA_real_,
      fixed_assets = NA_real_,

      ## depreciation (not available pre-1996)
      depr_land = NA_real_,
      depr_bldg = NA_real_,
      depr_lease = NA_real_,
      depr_fixed_equip = NA_real_,
      depr_auto = NA_real_,
      depr_major_equip = NA_real_,
      depr_minor_equip = NA_real_,

      ## balance sheet (not available pre-1996)
      current_assets = NA_real_,
      current_liabilities = NA_real_,

      ## PPS-equivalent variables (only populated in v1996 HCRIS)
      pps_ip_charges = NA_real_,
      pps_op_charges = NA_real_,
      pps_mcare_cost = NA_real_,
      pps_pgm_cost = NA_real_
    )

  assign(paste0("final.reports.", i), final.reports)
  if (i == 1985) {
    final.hcris.pps.early <- final.reports
  } else {
    final.hcris.pps.early <- rbind(final.hcris.pps.early, final.reports)
  }
}


# Combine all PPS data -----------------------------------------------------

final.hcris.pps <- rbind(final.hcris.pps.early, final.hcris.pps96)


# Write output -------------------------------------------------------------

write_tsv(final.hcris.pps, 'data/output/HCRIS_Data_PPS.txt', append=FALSE, col_names=TRUE)

File: data-code-py/H3_HCRIS_PPS.py
"""PPS (Prospective Payment System) extraction — 1985-1999.

Pre-extracted Stata flat files from NBER, using Form 2552-85/89/92/96.
- 1985-1995: limited variable set (no balance sheet, no revenue stmt)
- 1996-1999: full variable set, split into two parts per year
Date formats:
- Pre-1996: Julian YYDDD (e.g., "91274" = Oct 1, 1991)
- 1996-1999: CCYYMMDD (e.g., "19971001")
Output dates formatted as M/D/YYYY to match HCRIS H1/H2 output.
"""

from pathlib import Path
import numpy as np
import pandas as pd


# SSA state code mapping ---------------------------------------------------

SSA_STATE = pd.read_csv(
    Path("data-code/pps-xwalk/ssa_state_codes.csv"),
    dtype={"ssa_code": str, "state": str},
)
_STATE_MAP = dict(zip(SSA_STATE["ssa_code"], SSA_STATE["state"]))


# Helper functions ---------------------------------------------------------

def pad_provider(s):
    """Pad provider number to 6 digits with leading zeros."""
    s = pd.Series(s).astype(str).str.strip()
    return s.where(s != "", other=pd.NA).str.zfill(6)


def get_state_from_provider(provider_num):
    """Derive state abbreviation from first 2 digits of provider number."""
    return pd.Series(provider_num).astype(str).str[:2].map(_STATE_MAP)


def _strftime_mdy(dates):
    """Format datetime Series as M/D/YYYY (no leading zeros), cross-platform."""
    m = dates.dt.month.astype("Int64").astype(str)
    d = dates.dt.day.astype("Int64").astype(str)
    y = dates.dt.year.astype("Int64").astype(str)
    result = m + "/" + d + "/" + y
    return result.where(dates.notna(), other=pd.NA)


def parse_julian_date(date_str):
    """Parse Julian YYDDD strings to M/D/YYYY."""
    s = pd.Series(date_str).astype(str).str.strip()
    mask = s.notna() & (s.str.len() >= 5)
    yy = pd.to_numeric(s.str[:2], errors="coerce")
    ddd = pd.to_numeric(s.str[2:5], errors="coerce")
    year = np.where(yy >= 80, 1900 + yy, 2000 + yy)
    base = pd.to_datetime(
        pd.Series(year, dtype="Int64").astype(str) + "-01-01",
        errors="coerce",
    )
    dates = base + pd.to_timedelta(ddd - 1, unit="D")
    result = _strftime_mdy(dates)
    result = result.where(mask, other=pd.NA)
    return result


def parse_ccyymmdd(date_str):
    """Parse CCYYMMDD strings to M/D/YYYY."""
    s = pd.Series(date_str).astype(str).str.strip()
    mask = s.notna() & (s.str.len() == 8)
    dates = pd.to_datetime(s, format="%Y%m%d", errors="coerce")
    result = _strftime_mdy(dates)
    result = result.where(mask, other=pd.NA)
    return result


# Column template (consistent output order) --------------------------------

OUTPUT_COLS = [
    "report", "provider_number", "npi", "fy_start", "fy_end",
    "date_processed", "date_created", "status", "year", "data_source",
    "beds", "tot_charges", "net_pat_rev", "tot_discounts", "tot_operating_exp",
    "ip_charges", "icu_charges", "ancillary_charges",
    "tot_discharges", "mcare_discharges", "mcaid_discharges",
    "tot_mcare_payment", "secondary_mcare_payment",
    "street", "city", "state", "zip", "county", "name",
    "uncomp_care", "cost_to_charge",
    "new_cap_ass", "cash", "fixed_assets",
    "depr_land", "depr_bldg", "depr_lease", "depr_fixed_equip",
    "depr_auto", "depr_major_equip", "depr_minor_equip",
    "current_assets", "current_liabilities",
    "pps_ip_charges", "pps_op_charges", "pps_mcare_cost", "pps_pgm_cost",
]


# Import PPS 1996-1999 ----------------------------------------------------

def extract_pps_1996_1999():
    frames = []
    for yr in range(1996, 2000):
        yy = str(yr)[2:]
        base = Path("data/input/hcris_pps_nber")
        part1 = pd.read_stata(base / f"pps{yy}_f1_f1800.dta")
        part2 = pd.read_stata(base / f"pps{yy}_f1801_to_end.dta")
        part1["f1"] = pad_provider(part1["f1"]).values
        part2["f1"] = pad_provider(part2["f1"]).values
        pps = part1.merge(part2, on="f1", how="left")

        df = pd.DataFrame({
            "report": pd.NA,
            "provider_number": pps["f1"],
            "npi": pd.NA,
            "fy_start": parse_ccyymmdd(pps["f22"]).values,
            "fy_end": parse_ccyymmdd(pps["f23"]).values,
            "date_processed": pd.NA,
            "date_created": parse_ccyymmdd(pps["f69"]).values,
            "status": pps["f74"].astype(str),
            "year": yr,
            "data_source": "pps",
            "beds": pd.to_numeric(pps["f85"], errors="coerce"),
            "tot_charges": pd.to_numeric(pps["f2135"], errors="coerce"),
            "net_pat_rev": pd.to_numeric(pps["f2137"], errors="coerce"),
            "tot_discounts": pd.to_numeric(pps["f2136"], errors="coerce"),
            "tot_operating_exp": pd.to_numeric(pps["f2138"], errors="coerce"),
            "ip_charges": pd.to_numeric(pps["f2133"], errors="coerce"),
            "icu_charges": pd.to_numeric(pps["f2108"], errors="coerce"),
            "ancillary_charges": pd.to_numeric(pps["f2115"], errors="coerce"),
            "tot_discharges": pd.to_numeric(pps["f178"], errors="coerce"),
            "mcare_discharges": pd.to_numeric(pps["f172"], errors="coerce"),
            "mcaid_discharges": pd.to_numeric(pps["f175"], errors="coerce"),
            "tot_mcare_payment": pd.to_numeric(pps["f1833"], errors="coerce"),
            "secondary_mcare_payment": pd.to_numeric(pps["f1834"], errors="coerce"),
            "street": pd.NA,
            "city": pd.NA,
            "state": get_state_from_provider(pps["f1"]).values,
            "zip": pd.NA,
            "county": pd.NA,
            "name": pps["f18"].astype(str),
            "uncomp_care": np.nan,
            "cost_to_charge": np.nan,
            "new_cap_ass": pd.to_numeric(pps["f1473"], errors="coerce"),
            "cash": pd.to_numeric(pps["f2038"], errors="coerce"),
            "fixed_assets": pd.to_numeric(pps["f2066"], errors="coerce"),
            "depr_land": pd.to_numeric(pps["f2050"], errors="coerce"),
            "depr_bldg": pd.to_numeric(pps["f2054"], errors="coerce"),
            "depr_lease": pd.to_numeric(pps["f2056"], errors="coerce"),
            "depr_fixed_equip": pd.to_numeric(pps["f2058"], errors="coerce"),
            "depr_auto": pd.to_numeric(pps["f2060"], errors="coerce"),
            "depr_major_equip": pd.to_numeric(pps["f2062"], errors="coerce"),
            "depr_minor_equip": pd.to_numeric(pps["f2064"], errors="coerce"),
            "current_assets": pd.to_numeric(pps["f2048"], errors="coerce"),
            "current_liabilities": pd.to_numeric(pps["f2081"], errors="coerce"),
            "pps_ip_charges": np.nan,
            "pps_op_charges": np.nan,
            "pps_mcare_cost": np.nan,
            "pps_pgm_cost": np.nan,
        })
        frames.append(df[OUTPUT_COLS])
        print(f"  PPS {yr}: {len(df):,} reports")
    return pd.concat(frames, ignore_index=True)


# Import PPS 1985-1995 ----------------------------------------------------

def extract_pps_1985_1995():
    frames = []
    for yr in range(1985, 1996):
        yy = str(yr)[2:]
        pps = pd.read_stata(Path("data/input/hcris_pps_nber") / f"pps{yy}.dta")
        pps["provno"] = pad_provider(pps["provno"]).values

        df = pd.DataFrame({
            "report": pd.NA,
            "provider_number": pps["provno"],
            "npi": pd.NA,
            "fy_start": parse_julian_date(pps["begdate"]).values,
            "fy_end": parse_julian_date(pps["enddate"]).values,
            "date_processed": pd.NA,
            "date_created": pd.NA,
            "status": pps["status"].astype(str),
            "year": yr,
            "data_source": "pps",
            "beds": pd.to_numeric(pps["totbeds"], errors="coerce"),
            "tot_charges": pd.to_numeric(pps["totalg"], errors="coerce"),
            "net_pat_rev": np.nan,
            "tot_discounts": np.nan,
            "tot_operating_exp": pd.to_numeric(pps["opertots"], errors="coerce"),
            "ip_charges": pd.to_numeric(pps["mtotalg"], errors="coerce"),
            "icu_charges": np.nan,
            "ancillary_charges": pd.to_numeric(pps["outpatg"], errors="coerce"),
            "tot_discharges": pd.to_numeric(pps["f88"], errors="coerce"),
            "mcare_discharges": pd.to_numeric(pps["f82"], errors="coerce"),
            "mcaid_discharges": pd.to_numeric(pps["f84"], errors="coerce"),
            "tot_mcare_payment": pd.to_numeric(pps["amtdue"], errors="coerce"),
            "secondary_mcare_payment": np.nan,
            "street": pd.NA,
            "city": pd.NA,
            "state": get_state_from_provider(pps["provno"]).values,
            "zip": pd.NA,
            "county": pd.NA,
            "name": pps["provname"].astype(str),
            "uncomp_care": np.nan,
            "cost_to_charge": np.nan,
            "new_cap_ass": np.nan,
            "cash": np.nan,
            "fixed_assets": np.nan,
            "depr_land": np.nan,
            "depr_bldg": np.nan,
            "depr_lease": np.nan,
            "depr_fixed_equip": np.nan,
            "depr_auto": np.nan,
            "depr_major_equip": np.nan,
            "depr_minor_equip": np.nan,
            "current_assets": np.nan,
            "current_liabilities": np.nan,
            "pps_ip_charges": np.nan,
            "pps_op_charges": np.nan,
            "pps_mcare_cost": np.nan,
            "pps_pgm_cost": np.nan,
        })
        frames.append(df[OUTPUT_COLS])
        print(f"  PPS {yr}: {len(df):,} reports")
    return pd.concat(frames, ignore_index=True)


# Main ---------------------------------------------------------------------

def extract_pps():
    print("Extracting PPS data (1985-1999)...")
    early = extract_pps_1985_1995()
    late = extract_pps_1996_1999()
    combined = pd.concat([early, late], ignore_index=True)
    combined.to_csv("data/output/HCRIS_Data_PPS.txt", sep="\t", index=False)
    print(f"  PPS total: {len(combined):,} rows written")
    return combined


if __name__ == "__main__":
    extract_pps()
